### PR TITLE
bremBadger implementation

### PR DESCRIPTION
**Pull request by wtj2021**
spec https://docs.google.com/document/d/14Ub20r21C7Sqt05SeMyjRX4Qr76PxpeWPLZpC91GYHw/edit#heading=h.9cluwwsmy8dx

### CODE REVIEW

**Review by sajanrajdev — COMMENTED** [review]
## Logic Review
Specs implemented:
- Owner is allowed to enable deposits, this sets the deposit start and end automatically to 2 weeks - LG
- Deposits can be disabled by owner - LG
- shares are minted according to PPFS logic when user deposits - LG
- Introduces withdrawAll function that allows for the withdrawal of all vested shares up to that point
- Last week of vesting returns remaining shares for user - LG
- If user attempts to withdraw after the 12 weeks period is completed, they get their full shares withdrawn and the function returns 12 as the withdrawal period. - LG
- Includes view function to get vested amount in terms of underlying - LG

@wtj2021 the only thing missing is an owner only function to enable full withdrawals in case that the program is completed earlier (i.e. if breakeven is reached before the program's completion through a 4 twap on the BADGER price). In this case, governance will enable withdrawals atomically and stop emissions.

Outside of this, everything LGTM. Will review the tests.

File: src/bremBadger.sol [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {SafeERC20, IERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+contract bremBadger is ERC20Upgradeable, ReentrancyGuard, UUPSUpgradeable {
+    using SafeERC20 for IERC20;
+
+    // The start of the program is set retroactively to Feb 18, 2024. 
+    // This mean that the unlock must occur 9 months after this, on Nov 18, 2024.
```
Comment (sajanrajdev) [inline src/bremBadger.sol]: This _means_

**Review by sajanrajdev — COMMENTED** [review]
## Tests Review
Introduces the following dev env tests:
- Test deposit period and deposit process permissions - LG
- Tests PPFS math and donation handling: @wtj2021 would be good to see the ppfs working properly after the donation (i.e. a user withdrawing ang getting more/the donated amount). To ensure that the inclusion of this feature serves its purpose of preventing stuck assets.
- Tests that user's underlying vests properly on a weekly basis and withdraws all at the end of the 12 weeks, obtaining their full amount back - LG
- Tests partial withdrawals for a single user and tests a withdraw attempt with no shares - LG

@wtj2021 Current tests properly assess the base functionality and they look good. I think there's an opportunity to expand with the following edge/more realistic cases, where things could potentially break:
- Odd amounts - Using non round amounts could lead to revelaing potential accounting errors
- Having multiple users participating simultaneously.
- Testing against the actual fork of the contracts: Since this will be a one-off contract, I think it makes sense to test it with the real instance of remBADGER and its holders. A test like this could solve my two comments from above. There could be a lot of value on simulating the full cycle of a few potential participants depositing their full balances and withdrawing at difference cadences (weekly, by weekly, monthly, last week and a long time after the end of the program).

File: src/bremBadger.sol (hunk 1 above)
Comment (wtj2021) [reply src/bremBadger.sol]: fixed

**Review by sajanrajdev — APPROVED** [review]
## Changes Review
-  Removes transferability from vault to further simplify logic and mitigate high security findings
- Simplifies vesting math by using amount in and out tracking instead of shares math - LGTM
- Deposits can be enabled and disabled multiple times (good idea in case there are complains and there's an approval for people to deposit after the deposit period ends.
- Program termination is introduced via a flag. When triggered, withdrawals are enabled for all users for their full outstanding amount. Deposits are no longer allowed when the program is terminated (it cannot be reversed! - SGTM). - LG
-  View function for vesting amount modified accordingly.
- Stuck assets/donations would require an upgrade. Agree with the move, no need to overcomplicate the logic for this odd case. If it occurs, we can address as necessary.

Logic changes LGTM

## Tests Changes Review
- Fork and dev tests introduced
- Previous tests were adapted - LG
- Introduced a test for the interaction of multiple users (I modified the time warps locally so that moves are not to round weeks but rather odd times and the calculations worked as expected. In other words, claiming on week 5.9 after the unlock only allows claiming of 5 week's worth, as expected) - LG
- Tests that donations don't affect vesting estimations (donations lead to stuck funds).
- Tests program termination before and after unlock - LG
- Fork test suite tests deposit and withdrawal flows and full integration of flow of two users depositing all their balance and withdrawing multiple times throughout the program unlocking. - LGTM

Tests LGTM!

## Notes
LGTM

@wtj2021, I am not familiar with this proxy/upgradeability pattern but, if it is possible, could we make the Governance ProxyAdmin the upgrade authority for this:
- [devProxyAdmin](https://etherscan.io/address/0x20Dce41Acca85E8222D6861Aa6D23B6C941777bF#writeContract)

This contract can call `upgradeAndCall` or `upgrade` to any contract it is assigned to and is owned by the Governance 2 day [Timelock](https://etherscan.io/address/0x21CF9b77F88Adf8F8C98d7E33Fe601DC57bC0893). This would be the safest setup for managing upgrades.

As for the vault's owner, through discussions with @dapp-whisperer, we agreed the best actor to handle this is Badger's [TechOps](https://etherscan.io/address/0x86cbD0ce0c087b482782c181dA8d191De18C8275). It is agile enough and meets the security rigor that these operations require.

**Review by sajanrajdev — APPROVED** [review]
Introduced admin role for managing upgrades. Owner still controls the vault's flow. 

Upgrade test added for the fork proxyAdmin and it shows proper functioning. Test is passing. LGTM

**Review by GalloDaSballo — APPROVED** [review]
# QA - To Monitor

I would add these via upgrade only if necessary

If people end up depositing via donation
-> Add sweep via upgrade

If people end up forgetting to withdraw
-> Add a way to withdraw for them only once 100% of the program has expired via upgrade


# Changes
- Allows deposits
- Withdrawals are capped by maximum deposited - claimed
- Can only deposit until period or program ends
- Can withdraw only once time has passed or program has ended

**Pull request merged**